### PR TITLE
ARROW-3075: [C++] Merge parquet-cpp codebase into Arrow C++ codebase

### PR DESCRIPTION
**Pull request by wesm**
Merges the apache/parquet-cpp code tree into the apache/arrow C++ codebase for the community's combined development process.

### CODE REVIEW

**Comment by pitrou** [comment]
> Parquet has a lot of build flags; some of these will basically need to get redone. Because there are no tests for these build configurations, it's very hard to test all the permutations here.

Perhaps we should wait until someone shows up to complain about the removed build options?

File: cpp/cmake_modules/BuildUtils.cmake
Comment (pitrou) [inline cpp/cmake_modules/BuildUtils.cmake]: Hmm... wow. What is the difference? :-)

File: cpp/cmake_modules/BuildUtils.cmake
Comment (pitrou) [inline cpp/cmake_modules/BuildUtils.cmake]: What is the prefix for? Perhaps add a comment?

File: cpp/cmake_modules/BuildUtils.cmake
Comment (pitrou) [inline cpp/cmake_modules/BuildUtils.cmake]: Add a comment what this is about? (what is a test group?)

**Review by pitrou — COMMENTED** [review]
I suppose this is isn't finished (e.g. the CI changes are missing), but here are a few comments on the CMake changes.

File: cpp/cmake_modules/BuildUtils.cmake
Comment (wesm) [reply cpp/cmake_modules/BuildUtils.cmake]: I was testing out something and left this artifact. Will remove

File: cpp/cmake_modules/BuildUtils.cmake
Comment (wesm) [reply cpp/cmake_modules/BuildUtils.cmake]: OK

File: cpp/cmake_modules/BuildUtils.cmake
Comment (wesm) [reply cpp/cmake_modules/BuildUtils.cmake]: OK

**Comment by wesm** [comment]
@xhochy I stand awaiting instructions about how to proceed with the merge. My preference is to squash the history and, while we are at it, move the data files to http://github.com/apache/arrow-testing.

According to https://stackoverflow.com/a/15802324 through some trickery we can graft the src/parquet commit history onto the Arrow repo to be able to do an `--ff-merge`. If we are collectively OK with the idea of pulling ~400 commits into the Arrow repo then we could do this.

If we pull in the commit history with `filter-branch`, one question is what to do with the early "unclean" commits (no JIRAs) 

I can't claim to speak for @nongli but I suspect he wouldn't mind if we squashed these pre-Apache commits as the seed of the project. 

```
* 592cf71 2015-04-02 | PARQUET-232: minor compilation issue [Fabrizio Fabbri]
* db20bae 2014-09-08 | Add "parquet_reader.cc" in the folder "example". [Yue Chen]
* bce89a0 2014-10-28 | PARQUET-120: Copy dev scripts and readme from parquet-mr. [Nong Li]
* f23be56 2014-06-04 | Fix broken download_thirdparty script. [Nong Li]
* 153f92b 2014-06-03 | Add apache license. [Nong Li]
* 09d3c74 2014-06-02 | Switch to int64. [Nong Li]
* bfe23da 2014-06-02 | Add lz4 codec. [Nong Li]
* aa09914 2014-06-02 | Added a quick snappy & plain benchmark. [Nong Li]
* 143a868 2014-05-31 | Add thrift generated sources. [Nong Li]
* df3c37a 2014-05-31 | Implement snappy decompression. [Nong Li]
* dedad16 2014-05-31 | Abstraction for compression. [Nong Li]
* 0da0397 2014-05-31 | Move encoding to own folder. [Nong Li]
* 9a9125f 2014-05-31 | Add a plain encoded test file and some misc documentation. [Nong Li]
* e8be46d 2014-05-31 | Updated readme. [Nong Li]
* ef57939 2014-05-31 | Fix off by one in delta bit pack encoding. [Nong Li]
* 48591fd 2014-05-30 | fix core-dump while reading PLAIN_ENCODING; fix readNewPage to return page instead of the last one [Jaguar Xiong]
* 846e6b4 2014-05-27 | Fix max encoded size estimate in RLE encoding. [Nong Li]
* 46f5b52 2014-05-23 | Update readme. [Nong Li]
* 104373f 2014-05-23 | Move encodings to separate files. [Nong Li]
* 078bb78 2014-05-23 | Implement delta byte length and delta string encodings. [Nong Li]
* e214cb4 2014-05-22 | Fix delta binary encoding. Decodes at about ~200M/sec. [Nong Li]
* c656554 2014-05-22 | Cleanup bitpacking encoding. Add delta length encoding. [Nong Li]
* 992f187 2014-05-20 | Initial implementation of delta binary packing. [Nong Li]
* 722c751 2014-05-19 | Added simple decode benchmark. [Nong Li]
* 0959d2a 2014-05-19 | Inline some simple functions. [Nong Li]
* c23d93a 2014-05-19 | Change decoders to a batched API. [Nong Li]
* bc84379 2014-05-18 | Remove makefile. [Nong Li]
* 0ac13db 2014-05-18 | Update readme. [Nong Li]
* c19c5a7 2014-05-18 | Plumb through all the types. [Nong Li]
* 060c0d7 2014-05-18 | Better decoder management. [Nong Li]
* 9f1d702 2014-05-17 | Fix stream abstraction. [Nong Li]
* 674a392 2014-05-13 | Read a dictionary encoded int column. [Nong Li]
* f979b15 2014-05-13 | Implement some encodings. [Nong Li]
* f1b987e 2014-05-13 | Include some Impala utility code. [Nong Li]
* 9b42064 2014-05-13 | Can read the dictionary for ints. [Nong Li]
* eb10a21 2014-05-12 | Move build to cmake. [Nong Li]
* 01c30aa 2014-05-12 | Initial classes. [Nong Li]
* 08acdf6 2014-05-12 | Initial commit [Nong Li]
```

**Comment by wesm** [comment]
Here is the scrubbed, moved source directory:

https://github.com/wesm/parquet-cpp-merging

Here are the commands I used to create it:

```
# Remove everything except src/parquet
git filter-branch -f --tree-filter "find . -not -path './.git' -not -path './.git/*' -not -path './src' -not -path './src/parquet' -not -path './src/parquet/*' -delete" --prune-empty

# Move src/parquet to cpp/src/parquet
git filter-branch -f --tree-filter 'mkdir -p cpp/src && test -d src/parquet && mv src/parquet cpp/src/parquet || echo "Nothing to do"' HEAD
```

Still want to squash the remaining early commits. I will rebase these commits and Arrow master now and push them here

**Comment by wesm** [comment]
A commingled git history would be extremely messy to reason about, and make git bisect difficult. We have commits in both repos spread out over a ~3 year period

File: cpp/cmake_modules/ThirdpartyToolchain.cmake
Comment (pitrou) [inline cpp/cmake_modules/ThirdpartyToolchain.cmake]: Seems like there's a conflict remnant here (and below).

**Comment by kszucs** [comment]
Even with a single merge commit? It'd keep all the hashes and references.

**Comment by kszucs** [comment]
Merging parquet-cpp/master increases arrow's size by ~4MB. Filter branch indeed rewrites the hashes, but simply moving the files doesn't. The most straightforward way would be `moving -> git merge --allow-unrelated-histories`, but You probably see the long term disadvantages of that better than I do.

File: cpp/cmake_modules/BuildUtils.cmake
Comment (xhochy) [reply cpp/cmake_modules/BuildUtils.cmake]: The difference is that the old version is the variable/variable-name and the new one is the content of the variable. In the case of `if` this is irrelevant but matters at many more places.

**Comment by wesm** [comment]
Personally I would prefer to squash the history. Being willing to apply ~334 patches to preserve "git blame" on the code is about as far as I'm willing to compromise

**Comment by pitrou** [comment]
My two cents:
* it is mandatory to be able to bisect without unduly pain
* being able to `git blame` the reasons for a specific change is very useful

The current approach (keeping a linear history for the parquet source subdirectory) seems to satisfy those two constraints. If that approach is undesirable I think a squashed history is the only other solution.

**Comment by xhochy** [comment]
I'm happy with the current approach, we can keep it like that.

**Comment by wesm** [comment]
Alright. I'll work on getting this into merge-ready shape. Will there be some more patches to apply from parquet-cpp after the release goes out?

**Comment by wesm** [comment]
I need to deal with the test data still. But you can check out the dev workflow for Parquet-only:

```
ninja parquet
ctest -L parquet
```

This builds only the Parquet libraries and unit tests, and runs them. You can run all unit tests with `-L unittest` or only the Parquet unit tests with `-L parquet`.

Unfortunately I have two failing tests, evidently it is caused by some build toolchain issue but I haven't figured it out yet

```shell
$ gdb debug/parquet-file-serialize-test
SNIP

[  PASSED  ] 80 tests.
*** Error in `/home/wesm/code/arrow/cpp/build/debug/parquet-file-serialize-test': corrupted size vs. prev_size: 0x0000000000d1f0d0 ***

Program received signal SIGABRT, Aborted.
0x00007ffff5b3ac37 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
56	../nptl/sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  0x00007ffff5b3ac37 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007ffff5b3e028 in __GI_abort () at abort.c:89
#2  0x00007ffff5b772a4 in __libc_message (do_abort=1, fmt=fmt@entry=0x7ffff5c89350 "*** Error in `%s': %s: 0x%s ***\n") at ../sysdeps/posix/libc_fatal.c:175
#3  0x00007ffff5b82cd7 in malloc_printerr (ptr=0xd1f0d0, str=0x7ffff5c85464 "corrupted size vs. prev_size", action=<optimized out>) at malloc.c:4998
#4  malloc_consolidate (av=av@entry=0x7ffff5ec6760 <main_arena>) at malloc.c:4159
#5  0x00007ffff5b8372d in _int_free (av=0x7ffff5ec6760 <main_arena>, p=<optimized out>, have_lock=0) at malloc.c:4059
#6  0x00007ffff6e1502b in boost::re_detail_106700::mem_block_cache::~mem_block_cache() () from /home/wesm/cpp-toolchain/lib/libboost_regex.so.1.67.0
#7  0x00007ffff5b4053a in __cxa_finalize (d=0x7ffff7073750) at cxa_finalize.c:56
#8  0x00007ffff6daac53 in __do_global_dtors_aux () from /home/wesm/cpp-toolchain/lib/libboost_regex.so.1.67.0
#9  0x00007fffffffbe00 in ?? ()
#10 0x00007ffff7dea7da in _dl_fini () at dl-fini.c:252
Backtrace stopped: frame did not save the PC
```

**Comment by wesm** [comment]
The link libraries is messed up (compared verbose linker command vs. parquet-cpp build), let me work a bit more and I should be able to fix

**Comment by wesm** [comment]
OK, that's fixed -- the statically-linked libraries in libarrow.so were being leaked to transitive dependencies. I guess it didn't cause problems until now (the Parquet unit tests link to libarrow.so). 

I'm planning to put the data files from https://github.com/apache/parquet-cpp/tree/master/data into https://github.com/apache/parquet-testing and pull that during CI to set the `$PARQUET_TEST_DATA` env variable. Does that sound right?

**Comment by pitrou** [comment]
> I'm planning to put the data files from https://github.com/apache/parquet-cpp/tree/master/data into https://github.com/apache/parquet-testing and pull that during CI to set the $PARQUET_TEST_DATA env variable. Does that sound right?

Apparently there are 40KB in that directory. Do we really need a separate repo for that?
(if yes, perhaps make it a git submodule? btw, having to set supplementary env vars to run tests is a nuisance IMHO)

**Comment by wesm** [comment]
> Apparently there are 40KB in that directory. 

True, but I'm trying to encourage the Parquet community to become more proactive about cross-implementation testing, so having a testing/compatibility repo is a good place to start

>  btw, having to set supplementary env vars to run tests is a nuisance IMHO

I agree -- is there some way around this / some other way for the unit test executables to find the test data?

**Comment by pitrou** [comment]
> I agree -- is there some way around this / some other way for the unit test executables to find the test data?

if we're using a git submodule, the test path should be fixed and doesn't need to be set in an env var.

**Comment by wesm** [comment]
So what we could do is set the env variable if it is not set in the build-support/run-test.sh script

**Comment by pitrou** [comment]
Hmm, I didn't know that script existed... It doesn't seem terribly useful to have scripts lying around in the source tree if they are essentially undocumented ;-)

**Comment by pitrou** [comment]
In any case, apparently `git rev-parse --show-toplevel` gives you the root directory of the git clone you're into, assuming your build dir is _inside_ the git clone (which I'm assuming is a common situation) ;-) Invoking that command from a C++ test isn't obvious, though...

**Comment by wesm** [comment]
That's the script that's invoked when you run `ctest`; basically it captures logs and attempts to symbolize asan output, etc. (though it may be slightly in disrepair right now)

**Comment by pitrou** [comment]
Ah, I was unaware that `ctest` invoked a script.

**Comment by wesm** [comment]
OK, I put the data files inhttps://github.com/apache/parquet-testing. I'm setting env variables for the time being, we can refine the user experience in a subsequent patch

**Comment by wesm** [comment]
Beyond getting the build to pass, what else would you like to see on this initial integration? As a last matter, I'm interested in preserving the _committer_ field for the applied patches (rather than showing me as the committer). maybe it's a bit esoteric but scripts exist to cherry pick commits while preserving both author and committer: https://github.com/dingram/git-scripts/blob/master/scripts/git-cherry-pick-with-committer. So I might just write a Python script to run "rebase" those commits by cherry picking one by one

**Comment by wesm** [comment]
So the Linux C++ build is running at ~40 minutes again since the Parquet tests under valgrind are pretty slow. We may want to split the C++ testing off into its own build and test Python on Linux in its own build in the matrix. Follow up work

**Comment by wesm** [comment]
Some pretty weird errors in https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/build/1.0.7699/job/p5pkepbyw4xu1vgp. I will fix the Linux/macOS issues tomorrow

**Comment by wesm** [comment]
OK, I have fixed the "committer" field so it now respects the person who merged the patch in apache/parquet-cpp:

```
$ git log --pretty="%h Author %ae Committer %ce"
96e038c0 Author wesm+git@apache.org Committer wesm+git@apache.org
b8c62ae5 Author deepak.majeti@hpe.com Committer uwe@apache.org
e57f6dcb Author Uwe.Korn@blue-yonder.com Committer uwe@apache.org
194c867b Author antoine@python.org Committer wesm+git@apache.org
5255ed20 Author Jacek.Pliszka@gmail.com Committer wesm+git@apache.org
ec507314 Author cjjnjust@gmail.com Committer wesm+git@apache.org
1275f659 Author antoine@python.org Committer wesm+git@apache.org
318263f8 Author wesm+git@apache.org Committer wesm+git@apache.org
5b71098e Author deepak.majeti@hpe.com Committer wes.mckinney@twosigma.com
3f147e31 Author cjjnjust@gmail.com Committer wes.mckinney@twosigma.com
3f92e3e5 Author gidon@il.ibm.com Committer deepak.majeti@hpe.com
d640e286 Author antoine@python.org Committer wesm+git@apache.org
0e6795f2 Author Uwe.Korn@blue-yonder.com Committer uwe@apache.org
1def702a Author gidon@il.ibm.com Committer wesm+git@apache.org
76f2b65e Author wesm+git@apache.org Committer wesm+git@apache.org
faf4ba76 Author robbieg@uber.com Committer uwe@apache.org
8b5318c3 Author Uwe.Korn@blue-yonder.com Committer uwe@apache.org
5e4ee787 Author wes.mckinney@twosigma.com Committer wes.mckinney@twosigma.com
090cf0d4 Author antoine@python.org Committer wes.mckinney@twosigma.com
8f0a089d Author antoine@python.org Committer uwe@apache.org
100a87bc Author p.hoch@celonis.com Committer deepak.majeti@hpe.com
26bf892b Author p.hoch@celonis.com Committer deepak.majeti@hpe.com
4cbf1744 Author deepak.majeti@hpe.com Committer deepak.majeti@hpe.com
f684a0aa Author deepak.majeti@hpe.com Committer uwe@apache.org
dbb3e0f0 Author antoine@python.org Committer uwe@apache.org
...
```

**Comment by wesm** [comment]
It seems like the MSVC build flags and other configuration are different in Arrow vs. parquet-cpp. I'm seeing warnings like

```
C:\Miniconda36-x64\Scripts\clcache.exe  /nologo /TP -DARROW_EXPORTING -DBOOST_ALL_NO_LIB -D_CRT_SECURE_NO_WARNINGS -Dparquet_shared_EXPORTS -Isrc -I..\src -IC:\Miniconda36-x64\envs\arrow\Library\include -Igoogletest_ep-prefix\src\googletest_ep\include -Igflags_ep-prefix\src\gflags_ep\include -Isrc\rapidjson_ep\include -Iflatbuffers_ep-prefix\src\flatbuffers_ep-install\include -I..\thirdparty\hadoop\include -Izlib_ep\src\zlib_ep-install\include -Isnappy_ep\src\snappy_ep-install\include -Ibrotli_ep\src\brotli_ep-install\include -Ilz4_ep-prefix\src\lz4_ep\lib -Izstd_ep-prefix\src\zstd_ep\lib -Ithrift_ep\src\thrift_ep-install\include -Ithrift_ep\src\thrift_ep-install\include\thrift /DWIN32 /D_WINDOWS  /GR /EHsc /D_SILENCE_TR1_NAMESPACE_DEPRECATION_WARNING   /EHsc /bigobj /W3 /wd4800 /WX /MP /MD /O2 /Ob2 /DNDEBUG /showIncludes /Fosrc\parquet\CMakeFiles\parquet_shared.dir\arrow\writer.cc.obj /Fdsrc\parquet\CMakeFiles\parquet_shared.dir\ /FS -c ..\src\parquet\arrow\writer.cc
..\src\parquet/util/comparison.h(101): error C2220: warning treated as error - no 'object' file generated
..\src\parquet/util/comparison.h(101): warning C4910: 'parquet::CompareDefault<parquet::BooleanType>': '__declspec(dllexport)' and 'extern' are incompatible on an explicit instantiation
..\src\parquet/util/comparison.h(102): warning C4910: 'parquet::CompareDefault<parquet::Int32Type>': '__declspec(dllexport)' and 'extern' are incompatible on an explicit instantiation
..\src\parquet/util/comparison.h(103): warning C4910: 'parquet::CompareDefault<parquet::Int64Type>': '__declspec(dllexport)' and 'extern' are incompatible on an explicit instantiation
..\src\parquet/util/comparison.h(105): warning C4910: 'parquet::CompareDefault<parquet::FloatType>': '__declspec(dllexport)' and 'extern' are incompatible on an explicit instantiation
..\src\parquet/util/comparison.h(106): warning C4910: 'parquet::CompareDefault<parquet::DoubleType>': '__declspec(dllexport)' and 'extern' are incompatible on an explicit instantiation
```

these don't seem to be present in the Appveyor build in apache/parquet-cpp. Here is the compilation call there this code is compiled :

```
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\x86_amd64\CL.exe /c /I"C:\projects\parquet-cpp\build\src" /I"C:\projects\parquet-cpp\src" /IC:\Libraries\boost_1_63_0 /I"C:\projects\parquet-cpp\build\thrift_ep\src\thrift_ep-install\include" /I"C:\projects\parquet-cpp\build\thrift_ep\src\thrift_ep-install\include\thrift" /I"C:\projects\parquet-cpp\build\googletest_ep-prefix\src\googletest_ep\include" /I"C:\projects\parquet-cpp\build\release\include" /I"C:\Tools\vcpkg\installed\x64-windows\include" /nologo /W3 /WX /MP /O2 /Ob2 /D WIN32 /D _WINDOWS /D NDEBUG /D _CRT_SECURE_NO_WARNINGS /D PARQUET_EXPORTING /D BOOST_ALL_NO_LIB /D HAVE_INTTYPES_H /D HAVE_NETDB_H /D NOMINMAX /D "CMAKE_INTDIR=\"Release\"" /D _MBCS /Gm- /EHsc /MT /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /GR /Fo"parquet_objlib.dir\Release\\" /Fd"parquet_objlib.dir\Release\parquet_objlib.pdb" /Gd /TP /wd4996 /errorReport:queue  /bigobj "C:\projects\parquet-cpp\src\parquet\arrow\reader.cc" "C:\projects\parquet-cpp\src\parquet\arrow\record_reader.cc" "C:\projects\parquet-cpp\src\parquet\arrow\writer.cc" "C:\projects\parquet-cpp\src\parquet\bloom_filter.cc" "C:\projects\parquet-cpp\src\parquet\column_reader.cc" "C:\projects\parquet-cpp\src\parquet\column_scanner.cc" "C:\projects\parquet-cpp\src\parquet\column_writer.cc" "C:\projects\parquet-cpp\src\parquet\exception.cc" "C:\projects\parquet-cpp\src\parquet\file_reader.cc" "C:\projects\parquet-cpp\src\parquet\file_writer.cc" "C:\projects\parquet-cpp\src\parquet\metadata.cc" "C:\projects\parquet-cpp\src\parquet\murmur3.cc" "C:\projects\parquet-cpp\build\src\parquet\parquet_constants.cpp" "C:\projects\parquet-cpp\build\src\parquet\parquet_types.cpp" "C:\projects\parquet-cpp\src\parquet\printer.cc" "C:\projects\parquet-cpp\src\parquet\statistics.cc" "C:\projects\parquet-cpp\src\parquet\types.cc" "C:\projects\parquet-cpp\src\parquet\util\comparison.cc" "C:\projects\parquet-cpp\src\parquet\util\memory.cc"
```

**Comment by wesm** [comment]
Ugh I'm really hitting a wall on this C4910 issue -- I'm completel mystified why this has suddenly come up here and not in parquet-cpp. If anyone has a chance to take a look I would really appreciate it. I'll have access to a Windows machine later today which will help with iteration time

**Comment by wesm** [comment]
LOL, I missed

https://github.com/apache/parquet-cpp/blob/master/src/parquet/util/visibility.h#L24

I'm going to disable C4910 locally at the extern template usages in Parquet and open a JIRA to investigate how to not have the warning

**Comment by pitrou** [comment]
Note that Arrow already has a similar mechanism, see use of `ARROW_EXTERN_TEMPLATE` in `array.h`.

**Comment by wesm** [comment]
Yep, I know, this was one of the duplicated code bits between the two projects. The Parquet one accumulated some warning suppressions

**Comment by wesm** [comment]
I have access to a Windows machine finally today so I should be able to get the build passing...

**Comment by wesm** [comment]
To explain what's going on for people watching -- I ran into an issue where we need to use `__declspec(dllimport)` when linking Arrow into Parquet so that certain symbols are visible there (e.g. static members, like stuff in `arrow::CpuInfo`). We weren't using `dllimport` in Parquet (which could result in linker issues for people using parquet.dll) so I went ahead and added that here

**Comment by wesm** [comment]
Oof. I'll resolve the rebase conflict with pyarrow/parquet.py right before we merge this to make things less painful for me

**Comment by wesm** [comment]
I'm stuck on this linker issue:

```
   Creating library release\parquet-column_scanner-test.lib and object release\parquet-column_scanner-test.exp
column_scanner-test.cc.obj : error LNK2019: unresolved external symbol "private: static bool arrow::CpuInfo::initialized_" (?initialized_@CpuInfo@arrow@@0_NA) referenced in function "public: __cdecl parquet::DictEncoder<struct parquet::DataType<1> >::DictEncoder<struct parquet::DataType<1> >(class parquet::ColumnDescriptor const *,class parquet::ChunkedAllocator *,class arrow::MemoryPool *)" (??0?$DictEncoder@U?$DataType@$00@parquet@@@parquet@@QEAA@PEBVColumnDescriptor@1@PEAVChunkedAllocator@1@PEAVMemoryPool@arrow@@@Z)
parquet_static.lib(column_writer.cc.obj) : error LNK2001: unresolved external symbol "private: static bool arrow::CpuInfo::initialized_" (?initialized_@CpuInfo@arrow@@0_NA)
release\parquet-column_scanner-test.exe : fatal error LNK1120: 1 unresolved externals
[337/364] Building CXX object src\parquet\CMakeFiles\parquet-column_writer-test.dir\column_writer-test.cc.obj
ninja: build stopped: subcommand failed.
(arrow) C:\projects\arrow\cpp\build>set lastexitcode=1 
```

I haven't been able to figure out what is different about the linker configuration yet that is causing this to fail.

**Comment by wesm** [comment]
I plan to have this rebased and ready to merge today

**Comment by wesm** [comment]
OK, I have fixed the MSVC issues. I'm going to rebase and see if we get a clean build

**Comment by wesm** [comment]
pkgconfig for libparquet is currently broken so there are some issues with the Windows wheel build still. I will try to fix the pkgconfig later so we don't have to hack around it in the meantime.

**Comment by kou** [comment]
I think that the GLib build with Meson failure isn't a Meson problem.
Maybe, the Python environment that runs `/home/travis/miniconda/bin/meson` doesn't include Meson by `pip install meson` https://travis-ci.org/apache/arrow/jobs/425563611#L1493 .

The following debug print may help us:

```shell
which meson
head $(which meson)
$(head -n1 $(which meson) | sed -e 's/^#! *//') -c 'import os; import sys; import pprint; pprint.pprint([list(os.walk(path)) for path in sys.path])'
```
```

**Comment by wesm** [comment]
@kou I pinned meson to the prior version 0.47.1 in https://github.com/apache/arrow/pull/2507 and it fixed the build. I think it's a problem with meson version 0.47.2 released on August 25

**Comment by kou** [comment]
Umm. It's strange... Because https://travis-ci.org/apache/arrow/jobs/425583565#L3060 uses Meson 0.47.2 and there are no error.

Anyway, it's OK to me that we use 0.47.1 for now.

**Comment by wesm** [comment]
Just rebased this again. In theory should have a fully green build now

**Comment by wesm** [comment]
Just in case anyone gets an itchy trigger finger

* do NOT use `git rebase` with this branch -- I have been taking great care to preserve the $COMMITTER for each historical patch

* do NOT use the merge script

This should be merged to apache master with `git merge --ff-only`

**Review by xhochy — APPROVED** [review]
+1, LGTM

also verified that the build works locally.

**Comment by wesm** [comment]
Thanks. I'm just rebasing now; if you could hold off on merging any patches to master until I can FF-merge this that would be great, should be in 20-30 minutes

**Comment by wesm** [comment]
Build is clean (see my appveyor builds: https://ci.appveyor.com/project/wesm/arrow/build/1.0.2049), merging

**Pull request merged**
